### PR TITLE
Fast Pair BLE Device Discovery for Windows in Rust

### PR DESCRIPTION
**Pull request by TheShepord**
## Summary
Implementing Fast Pair BLE Device Discovery for Windows in Rust, as part of intern project. See go/fastpair-seeker-rust for further discussion on design decisions (Design > API Decisions).

## How did you test this change?
This is an in-development prototype, so unit and integration tests will be provided at a later date (b/288592509). 
`cargo run` output for Windows: 
image (1)

`cargo run` output for unsupported platform:

image (2)

### CODE REVIEW

File: fastpair/rust/src/bluetooth/mod.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+cfg_if::cfg_if! {
+    if #[cfg(windows)] {
+        mod windows_ble;
+        use windows_ble as imp;
+    } else {
+        compile_error!("Unsupported target platform. Currently supported: Windows.");
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/mod.rs:22]: may be better to just panic here instead of compile_error just so github actions can pass. would you also mind adding a build rust on windows github action to test the new windows code?

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs [hunk 2]
```diff
@@ -35,6 +73,87 @@ impl BleAdapter {
 
         Ok(BleAdapter { inner })
     }
+
+    pub fn scanner(&self) -> Result<impl Stream<Item = BleDevice>, anyhow::Error> {
+        let watcher = BluetoothLEAdvertisementWatcher::new()?;
+        watcher.SetScanningMode(BluetoothLEScanningMode::Active)?;
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/windows_ble/adapter.rs:79]: this didn't work for me afaik, so maybe catching an error here and logging a warning instead may be a better approach.

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs [hunk 3]
```diff
@@ -35,6 +73,87 @@ impl BleAdapter {
...
+
+        if self.inner.IsExtendedAdvertisingSupported()? {
+            watcher.SetAllowExtendedAdvertisements(true)?;
+        }
+
+        let (sender, receiver) = futures::channel::mpsc::channel(16);
+        let sender = Arc::new(std::sync::Mutex::new(sender));
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/windows_ble/adapter.rs:86]: why is this needed?

Comment (anayw2001) [reply fastpair/rust/src/bluetooth/windows_ble/adapter.rs:86]: nvm

**Comment by TheShepord** [comment]
> may be better to just panic here instead of compile_error just so github actions can pass. would you also mind adding a build rust on windows github action to test the new windows code?

Added `unsupported` module, calling any methods cause a panic.


> this didn't work for me afaik, so maybe catching an error here and logging a warning instead may be a better approach. 

Unsure what you mean by this, could you elaborate?

> why is this needed?

Added comments to prevent further confusion.

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs (hunk 2 above)
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/windows_ble/adapter.rs:82]: Unsure what you mean by this, which part didn't work for you and which warning should I be logging?

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs (hunk 3 above)
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/windows_ble/adapter.rs:90]: Added comment to prevent further confusion.

**Comment by TheShepord** [comment]
Changed from wrapper to trait-based implementation for cross-platform support. General cleanup. @anayw2001, @suetfei.

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs (hunk 2 above)
Comment (anayw2001) [reply fastpair/rust/src/bluetooth/windows_ble/adapter.rs:87]: the SetScanningMode to be specific, apparently it's a hardware-dependent call? weird but yeah definitely log a warning if that was unable to be set, but i should imagine the applet would still work.

File: fastpair/rust/rustfmt.toml [hunk 4]
```diff
@@ -0,0 +1 @@
+max_width = 80
```
Comment (marshallpierce) [inline fastpair/rust/rustfmt.toml:1]: why so short?

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        // pinned box to satisfy trait bounds for `Stream`.
+        Ok(Box::pin(receiver.filter_map(move |event_args| {
+            //  Move `watcher` into `FilterMap` closure. This ensures `watcher`
+            // is only dropped when the stream is closed.
+            let _watcher = &watcher;
+
+            // Move `event_args` into async block.
+            async move {
+                match event_args.AdvertisementType().ok()? {
+                    BluetoothLEAdvertisementType::NonConnectableUndirected => None,
```
Comment (marshallpierce) [inline fastpair/rust/src/bluetooth/windows_ble/adapter.rs:154]: Why are these skipped?

**Comment by marshallpierce** [comment]
Also, I suggest running `rustfmt` and `cargo clippy --all-targets` to get hints on stuff to clean up

File: fastpair/rust/src/bluetooth/common.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    async fn default() -> Result<Box<Self>, anyhow::Error>;
+
+    // Scan for nearby devices, returning a `Stream` of futures that can be
+    // iterated over and polled to retrieve `BleDevice`.
+    //
+    // NOTE: Using Boxed dyn here is silly because in cross-platform code there
+    // should only ever be one concrete type implementing Adapter. Change this
+    // to `impl Stream` once impl trait return types are stabilized in traits.
+    // b/289224233.
+    fn scan_devices(&self) -> Result<Pin<Box<dyn Stream<Item = Self::TDevice>>>, anyhow::Error>;
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/common.rs:38]: could we do something like next_device(&self) and have it block until a device is seen? that way we can put it in a nice loop {} block.

File: fastpair/rust/src/bluetooth/common.rs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// Concrete types implementing this trait represent Bluetooth Peripheral devices.
+// They provide methods for retrieving device info and running device actions,
+// such as pairing.
+pub trait Device {
+    // Retrieve the name advertised by this device.
+    fn name(&self) -> Result<String, anyhow::Error>;
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/common.rs:46]: maybe getting mac address?

File: fastpair/rust/src/bluetooth/unsupported/adapter.rs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+use async_trait::async_trait;
+use futures::stream::Stream;
+
+use super::BleDevice;
+use crate::bluetooth::common::Adapter;
+
+// Concrete type implementing `Adapter`, used for unsupported devices.
+// Every method should panic.
+pub struct BleAdapter {}
```
Comment (anayw2001) [inline fastpair/rust/src/bluetooth/unsupported/adapter.rs:25]: can just be pub struct BleAdapter;

File: fastpair/rust/rustfmt.toml (hunk 4 above)
Comment (anayw2001) [reply fastpair/rust/rustfmt.toml:1]: g3 rustfmt will reformat these if saved in cider or something so that's probably why.

File: fastpair/rust/src/bluetooth/windows_ble/adapter.rs (hunk 5 above)
Comment (anayw2001) [reply fastpair/rust/src/bluetooth/windows_ble/adapter.rs:154]: i suspect fastpair doesn't need these, but i will probably go through this at some point later and make an enum for the types to be exported back to the caller.

File: fastpair/rust/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Library file, exports modules for use in integration tests and external crates.
+
+pub mod bluetooth;
```
[the remark this answers is not in the document]
Comment (suetfei) [reply fastpair/rust/src/lib.rs:17]: They are not used now, but will be later for Fastpair V2 or whatever.

File: fastpair/rust/src/main.rs [hunk 10]
```diff
@@ -12,9 +12,22 @@
...
-mod message_stream;
-mod types;
+mod bluetooth;
 
-fn main() {
-    println!("Fastpair Rust!");
+use futures::executor;
+use futures::stream::StreamExt;
+
+fn main() -> Result<(), anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/main.rs:22]: For now I'll leave it here since I'm developing a scanner demo as my intern project. The most likely restructure afterwards is making bluetooth a standalone top-level crate alongside a separate scanner crate. Thank you!

File: fastpair/rust/src/bluetooth/windows_ble/device.rs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+impl BleDevice {
+    pub async fn from_addr(
+        addr: u64,
+        kind: BluetoothAddressType,
+    ) -> Result<Self, anyhow::Error> {
+        let inner =
+            BluetoothLEDevice::FromBluetoothAddressWithBluetoothAddressTypeAsync(addr, kind)?
+                .await?;
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/windows_ble/device.rs:41]: I'm against making this a tuple struct to facilitate adding more parameters in the future. Map here doesn't quite work unfortunately, error types are mismatched. I'll fix up the error types as per your other comment, let me know if the new design works.

File: fastpair/rust/src/bluetooth/windows_ble/device.rs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    BluetoothLEDevice,
+};
+
+pub struct BleDevice {
+    inner: BluetoothLEDevice,
+}
+
+impl BleDevice {
+    pub async fn from_addr(
+        addr: u64,
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/windows_ble/device.rs:36]: I'll leave as-is for now since this is a private method and Windows bluetooth addresses use u64. Will make a note for future.

File: fastpair/rust/src/bluetooth/mod.rs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod common;
+
+cfg_if::cfg_if! {
+    if #[cfg(windows)] {
+        mod windows_ble;
+        pub use windows_ble::*;
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/mod.rs:20]: To be clear, are you suggesting a top-level wrapper struct that calls platform-specific structs? Or rather, are you suggesting explicitly exporting every struct (e.g. pub use windows_ble::BleAdapter as BleAdapter)?

File: fastpair/rust/Cargo.toml [hunk 14]
```diff
@@ -20,3 +20,15 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+anyhow = "1.0"
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/Cargo.toml:23]: I'll file a bug and fix the error types once the bluetooth module is split into its own crate.

File: fastpair/rust/src/bluetooth/common.rs (hunk 6 above)
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/common.rs:38]: I'm fairly certain this is already the case? Stream provides poll_next, so we can do stream.next().await as shown in main.rs. Is this not what you mean?

File: fastpair/rust/src/bluetooth/common.rs (hunk 7 above)
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/common.rs:46]: I'll add additional calls as needed, there's a lot of information that could be nice to fetch.

File: fastpair/rust/src/bluetooth/common.rs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+use async_trait::async_trait;
+use futures::stream::Stream;
+
+// Concrete types implementing this trait are Bluetooth Central devices.
+// They provide methods for retrieving nearby connections and device info.
+#[async_trait]
+// Concrete types implementing this trait are Bluetooth Central devices.
+// They provide methods for retrieving nearby connections and device info.
+pub trait Adapter {
+    type TDevice: Device;
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/common.rs:26]: I've seen this employed in libraries as a shorthand for "Type." Happy to change if this isn't actually a thing.

File: fastpair/rust/src/bluetooth/common.rs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+    // Retrieve the system-default Bluetooth adapter.
+    async fn default() -> Result<Box<Self>, anyhow::Error>;
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/common.rs:29]: Filed a bug, I'll get around to this once my project demo is finalized.

File: fastpair/rust/src/bluetooth/unsupported/mod.rs [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Bluetooth LE module for unsupported devices. Every method panics.
+
+mod adapter;
```
[the remark this answers is not in the document]
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/unsupported/mod.rs:17]: To be clear, you're suggesting I make the unsupported module a single unsupported.rs file as opposed to a folder?

File: fastpair/rust/src/bluetooth/common.rs (hunk 6 above)
Comment (anayw2001) [reply fastpair/rust/src/bluetooth/common.rs:36]: just in a slight effort to make this return type slightly less ugly tbh

**Review by anayw2001 — APPROVED** [review]
lgtm! stream type is annoying, but we'll figure it out _eventually_.

File: fastpair/rust/src/bluetooth/common.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (marshallpierce) [reply fastpair/rust/src/bluetooth/common.rs:21]: I don't see it in practice. Maybe some C# refugee developer who still bears the scars of previous suffering. Nowadays anyone who's using an IDE that isn't syntax highlighting associated types is just a masochist.

File: fastpair/rust/src/bluetooth/unsupported/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (marshallpierce) [reply fastpair/rust/src/bluetooth/unsupported/mod.rs:16]: Yep, that'd work.

File: fastpair/rust/src/bluetooth/unsupported/adapter.rs [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+/// Concrete type implementing `Adapter`, used for unsupported devices.
+/// Every method should panic.
+pub struct BleAdapter;
+
+#[async_trait]
+impl Adapter for BleAdapter {
+    type Device = BleDevice;
+
+    async fn default() -> Result<Self, anyhow::Error> {
+        panic!("Unsupported target platform.");
```
Comment (marshallpierce) [inline fastpair/rust/src/bluetooth/unsupported/adapter.rs:29]: there is also `unimplemented!` which does the same thing, and maybe is more communicative to the reader? Or not. I'm not sure which is better.

**Review by marshallpierce — APPROVED** [review]

File: fastpair/rust/src/bluetooth/unsupported/adapter.rs (hunk 18 above)
Comment (TheShepord) [reply fastpair/rust/src/bluetooth/unsupported/adapter.rs:29]: I personally prefer `panic!` here. If there's a specific platform with pending features then `unimplemented!` makes more sense, but for an entire unsupported platform what's unimplemented isn't the specific function, but rather the entire platform library.
